pkg/gadgets/trace/open/tracer/tracer_test.go

```diff
@@ -0,0 +1,213 @@
+//go:build linux
+// +build linux
+
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer_test
+
+import (
+	"fmt"
+	"testing"
+
+	"golang.org/x/sys/unix"
+
+	utilstest "github.com/kinvolk/inspektor-gadget/pkg/gadgets/internal/test"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/trace/open/tracer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/trace/open/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
+
+func TestCreateTracer(t *testing.T) {
+	utilstest.RequireRoot(t)
+
+	eventCallback := func(types.Event) {}
+
+	tracer, err := tracer.NewTracer(&tracer.Config{}, nil, eventCallback)
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	if tracer == nil {
+		t.Fatal("Returned tracer was nil")
+	}
+	t.Cleanup(func() { tracer.Stop() })
+}
+
+func TestStopTracerIdempotent(t *testing.T) {
+	utilstest.RequireRoot(t)
+
+	eventCallback := func(types.Event) {}
+
+	tracer, err := tracer.NewTracer(&tracer.Config{}, nil, eventCallback)
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	if tracer == nil {
+		t.Fatal("Returned tracer was nil")
+	}
+	t.Cleanup(func() { tracer.Stop() })
+
+	// Check that a double stop doesn't cause issues
+	tracer.Stop()
+	tracer.Stop()
+}
+
+func TestTracer(t *testing.T) {
+	utilstest.RequireRoot(t)
+
+	var fileFd int
```

Comment:
It's a very interesting point. I reworked the code to avoid using those magic variables. I changed the code to make `generateEvent` return something (catPid for exec tests, fileFd for open tests, etc.), then that value is passed down to the function that generates the expected event. Looks a more clear (and safer) in that way IMO.

Reply:
Looks better now, cool! Marking as resolved.